src/olympia/devhub/forms.py

```diff
@@ -288,12 +291,32 @@ def save(self, commit=True):
 class WithSourceMixin(object):
     def clean_source(self):
         source = self.cleaned_data.get('source')
-        if source and not source.name.endswith(VALID_SOURCE_EXTENSIONS):
-            raise forms.ValidationError(
-                ugettext(
-                    'Unsupported file type, please upload an archive '
-                    'file {extensions}.'.format(
-                        extensions=VALID_SOURCE_EXTENSIONS)))
+        if source:
+            try:
+                if source.name.endswith('.zip'):
+                    zip_file = SafeZip(source)
+                    # testzip() returns None if there are no broken CRCs.
+                    if zip_file.zip_file.testzip() is not None:
+                        raise zipfile.BadZipfile()
+                elif source.name.endswith(('.tar.gz', '.tar.bz2')):
```

Comment:
Could we do something that `VALID_SOURCE_EXTENSIONS` here instead?  Seems weird we have that constant but don't actually use to determine what is a valid extension.

Reply:
I thought something like `elif os.path.splitext(source.name)[1] in VALID_SOURCE_EXTENSIONS:` but yeah could as easily be removed entirely.

Reply:
`elif os.path.splitext(source.name)[1] in VALID_SOURCE_EXTENSIONS` doesn't separate `.tar.foo` from `.zip` though, and it'd be kinda wonky to do an explicit `.zip` check above and not below, so... yeah, best avoided I think. I'll add a commit to try to get rid of that constant entirely.

Reply:
Nope, I have to keep it, or something like that. We need the extension and `splitext` doesn't handle `.tar.gz` (how could it...) so we have to hardcode a check against .tar.gz somewhere. Might as well keep this then. It does have a comment next to the constant so hopefully that will be enough.